// angular2-tree-diagram/esm5/lib/classes/node.class.js
var TreeDiagramNode = /** @class */ (function () {
    function TreeDiagramNode(props, config, getThisNodeList) {
        this.getThisNodeList = getThisNodeList;
        if (!props.guid) {
            return;
        }
        for (var prop in props) {
            if (props.hasOwnProperty(prop)) {
                this[prop] = props[prop];
            }
        }
        this.toggleState = false;
        if (config.nodeWidth) {
            this.width = config.nodeWidth;
        }
        if (config.nodeHeight) {
            this.height = config.nodeHeight;
        }
        this.children = new Set(props.children);
    }
    Object.defineProperty(TreeDiagramNode.prototype, "isMaker", {
        get: function () {
            return false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeDiagramNode.prototype, "isExpanded", {
        get: function () {
            return this.toggleState;
        },
        enumerable: true,
        configurable: true
    });
    TreeDiagramNode.prototype.destroy = function () {
        this.getThisNodeList().destroy(this.guid);
    };
    TreeDiagramNode.prototype.hasChildren = function () {
        return !!this.children.size;
    };
    TreeDiagramNode.prototype.toggle = function (state) {
        if (state === void 0) { state = !this.toggleState; }
        this.toggleState = state;
        if (state) {
            this.getThisNodeList().toggleSiblings(this.guid);
        }
    };
    TreeDiagramNode.prototype.childrenCount = function () {
        return this.children.size;
    };
    TreeDiagramNode.prototype.isRoot = function () {
        return this.parentId == null;
    };
    TreeDiagramNode.prototype.dragenter = function (event) {
        event.dataTransfer.dropEffect = 'move';
    };
    TreeDiagramNode.prototype.dragleave = function (event) {
        this.isDragover = false;
    };
    TreeDiagramNode.prototype.dragstart = function (event) {
        event.dataTransfer.effectAllowed = 'move';
        this.isDragging = true;
        this.toggle(false);
        this.getThisNodeList().draggingNodeGuid = this.guid;
    };
    TreeDiagramNode.prototype.dragover = function (event) {
        event.preventDefault();
        if (!this.isDragging) {
            this.isDragover = true;
        }
        event.dataTransfer.dropEffect = 'move';
        return false;
    };
    TreeDiagramNode.prototype.dragend = function () {
        this.isDragover = false;
        this.isDragging = false;
    };
    TreeDiagramNode.prototype.drop = function (event) {
        event.preventDefault();
        var guid = this.getThisNodeList().draggingNodeGuid;
        this.getThisNodeList().transfer(guid, this.guid);
        return false;
    };
    TreeDiagramNode.prototype.addChild = function () {
        var newNodeGuid = this.getThisNodeList().newNode(this.guid);
        this.children.add(newNodeGuid);
        this.toggle(true);
    };
    return TreeDiagramNode;
}());
export { TreeDiagramNode };
