// angular2-tree-diagram/esm5/lib/classes/node-maker.class.js
import { __extends } from "tslib";
import { TreeDiagramNode } from './node.class';
var TreeDiagramNodeMaker = /** @class */ (function (_super) {
    __extends(TreeDiagramNodeMaker, _super);
    function TreeDiagramNodeMaker() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isMakerState = true;
        return _this;
    }
    Object.defineProperty(TreeDiagramNodeMaker.prototype, "isMaker", {
        get: function () {
            return this.isMakerState;
        },
        enumerable: true,
        configurable: true
    });
    TreeDiagramNodeMaker.prototype.drop = function (event) {
        event.preventDefault();
        var guid = this.getThisNodeList().draggingNodeGuid;
        this.getThisNodeList().rootNode(guid);
        this.displayName = 'New node';
        return false;
    };
    TreeDiagramNodeMaker.prototype.dragenter = function (event) {
        event.dataTransfer.dropEffect = 'move';
        var guid = this.getThisNodeList().draggingNodeGuid;
        var node = this.getThisNodeList().getNode(guid);
        if (node.parentId) {
            this.displayName = 'Root';
        }
    };
    TreeDiagramNodeMaker.prototype.dragover = function (event) {
        event.preventDefault();
        var guid = this.getThisNodeList().draggingNodeGuid;
        var node = this.getThisNodeList().getNode(guid);
        if (!this.isDragging && node.parentId) {
            this.isDragover = true;
            event.dataTransfer.dropEffect = 'move';
        }
        return false;
    };
    TreeDiagramNodeMaker.prototype.dragleave = function (event) {
        this.displayName = 'New node';
        this.isDragover = false;
    };
    return TreeDiagramNodeMaker;
}(TreeDiagramNode));
export { TreeDiagramNodeMaker };
