// angular2-tree-diagram/esm2015/lib/classes/node.class.js
export class TreeDiagramNode {
    constructor(props, config, getThisNodeList) {
        this.getThisNodeList = getThisNodeList;
        if (!props.guid) {
            return;
        }
        for (const prop in props) {
            if (props.hasOwnProperty(prop)) {
                this[prop] = props[prop];
            }
        }
        this.toggleState = false;
        if (config.nodeWidth) {
            this.width = config.nodeWidth;
        }
        if (config.nodeHeight) {
            this.height = config.nodeHeight;
        }
        this.children = new Set(props.children);
    }
    get isMaker() {
        return false;
    }
    get isExpanded() {
        return this.toggleState;
    }
    destroy() {
        this.getThisNodeList().destroy(this.guid);
    }
    hasChildren() {
        return !!this.children.size;
    }
    toggle(state = !this.toggleState) {
        this.toggleState = state;
        if (state) {
            this.getThisNodeList().toggleSiblings(this.guid);
        }
    }
    childrenCount() {
        return this.children.size;
    }
    isRoot() {
        return this.parentId == null;
    }
    dragenter(event) {
        event.dataTransfer.dropEffect = 'move';
    }
    dragleave(event) {
        this.isDragover = false;
    }
    dragstart(event) {
        event.dataTransfer.effectAllowed = 'move';
        this.isDragging = true;
        this.toggle(false);
        this.getThisNodeList().draggingNodeGuid = this.guid;
    }
    dragover(event) {
        event.preventDefault();
        if (!this.isDragging) {
            this.isDragover = true;
        }
        event.dataTransfer.dropEffect = 'move';
        return false;
    }
    dragend() {
        this.isDragover = false;
        this.isDragging = false;
    }
    drop(event) {
        event.preventDefault();
        const guid = this.getThisNodeList().draggingNodeGuid;
        this.getThisNodeList().transfer(guid, this.guid);
        return false;
    }
    addChild() {
        const newNodeGuid = this.getThisNodeList().newNode(this.guid);
        this.children.add(newNodeGuid);
        this.toggle(true);
    }
}
