// angular2-tree-diagram/esm2015/lib/tree.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { NodesListService } from './services/nodes-list.service';
let TreeComponent = class TreeComponent {
    constructor(nodesSrv, sanitizer) {
        this.nodesSrv = nodesSrv;
        this.sanitizer = sanitizer;
        this.config = {
            nodeWidth: 200,
            nodeHeight: 100
        };
        this.paneDragging = false;
        this.zoom = 1;
        this.paneX = 0;
        this.paneY = 0;
    }
    get paneTransform() {
        return this.paneTransformState;
    }
    set paneTransform(value) {
        this.paneTransformState = value;
    }
    set data(data) {
        if (!data || !Array.isArray(data.json)) {
            return;
        }
        if (typeof data.config === 'object') {
            this.config = Object.assign(this.config, data.config);
        }
        this.nodes = this.nodesSrv.loadNodes(data.json, this.config);
    }
    get nodeMaker() {
        return this.nodesSrv.makerNode();
    }
    newNode() {
        this.nodesSrv.newNode();
    }
    onmousedown() {
        this.paneDragging = true;
    }
    onmousemove(event) {
        if (this.paneDragging) {
            const { movementX, movementY } = event;
            this.paneX += movementX;
            this.paneY += movementY;
            this.makeTransform();
        }
    }
    onmouseup() {
        this.paneDragging = false;
    }
    makeTransform() {
        this.paneTransform = this.sanitizer.bypassSecurityTrustStyle(`translate(${this.paneX}px, ${this.paneY}px) scale(${this.zoom})`);
    }
    preventMouse(event) {
        event.stopPropagation();
    }
    onmousewheel(event) {
        let delta;
        event.preventDefault();
        delta = event.detail || event.wheelDelta;
        this.zoom += delta / 1000 / 2;
        this.zoom = Math.min(Math.max(this.zoom, 0.2), 3);
        this.makeTransform();
    }
};
TreeComponent.ctorParameters = () => [
    { type: NodesListService },
    { type: DomSanitizer }
];
__decorate([
    Input()
], TreeComponent.prototype, "data", null);
TreeComponent = __decorate([
    Component({
        // tslint:disable-next-line:component-selector
        selector: 'tree-diagram',
        template: "<div\n  class=\"tree-pane\"\n  (mousedown)=\"onmousedown()\"\n  (mouseup)=\"onmouseup()\"\n  (mousemove)=\"onmousemove($event)\"\n  (mousewheel)=\"onmousewheel($event)\"\n>\n  <div class=\"tree-paning-container\" [style.transform]=\"paneTransform\">\n    <div *ngIf=\"nodes\" class=\"tree-roots-elements\">\n\n      <tree-diagram-node\n        [nodeId]=\"node.guid\"\n        class=\"tree-root tree-node\"\n        (mousedown)=\"preventMouse($event)\"\n        *ngFor=\"let node of nodes.roots\"\n      ></tree-diagram-node>\n\n      <tree-diagram-node\n        [nodeId]=\"nodeMaker\"\n        (click)=\"newNode()\"\n        (mousedown)=\"preventMouse($event)\"\n        class=\"tree-root tree-new-node tree-node\"\n      ></tree-diagram-node>\n\n    </div>\n  </div>\n</div>\n",
        styles: [":host{-webkit-print-color-adjust:exact;position:relative;display:block;-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;overflow:hidden;height:100vh;text-align:center}.tree-roots-elements{position:relative;text-align:center;display:inline-block;white-space:nowrap;cursor:default!important;font-size:0;-webkit-transform-origin:center;transform-origin:center}.tree-node{position:relative;display:inline-block;margin:15px;vertical-align:top}.tree-node:only-of-type>.line-to{display:none}.tree-pane,.tree-paning-container{position:absolute;left:0;top:0;right:0;bottom:0}"]
    })
], TreeComponent);
export { TreeComponent };
